### ISSUE TITLE
Tagging for screenplay breakdown

### ISSUE THREAD

**Issue by lmparppei** [lmparppei/Beat#52]
During the spring, I'll look a bit into making it possible to tag the screenplay inside Beat.

This would work as simply as adding attributes to string ranges (such as `"Prop": @YES`) and saving them either as notes in the screenplay – or as ranges inside `BeatDocumentSettings`. The latter might be the safer solution for compatibility with other Fountain editors.

The screenplay should probably be locked for this to be possible (need to add a `BeatDocumentSettings` value for this), as otherwise the current formatting system would just override any attributes. It could be modified to keep the tagging attributes, though.

I wrote small screenplay plugin, and realized how easy this would be to implement and to export as CVS. If somebody was smart enough to figure out how to write `.sex` (Movie Magic Scheduling) files, the results could also be imported straight into MMS.

This is how we run through the tags:

```
    for (NSTextStorage *storage in self.textView.textStorage.attributeRuns) {
        NSDictionary *attrs = [storage attributesAtIndex:0 effectiveRange:nil];
        // Check if the attributes contain Prop, Character, VFX, etc. and add to report
    }
```

**Comment by lmparppei** [comment]
Tagging is now somewhat implemented and seems to work. It requires the user to lock the screenplay, though.

![image](https://user-images.githubusercontent.com/20226704/107155323-666bc000-6980-11eb-8951-9d95de22416e.png)

To provide compatibility with industry-standard scheduling apps, the tagged data should be written into FDX. Here the real fun begins. Fountain to FDX can already be a bit quirky at times, even without tagging. The old `FDXInterface` has to be reworked to use `Line` objects from the parser, so our tags (which are NOT contained in the Fountain markup but saved into `BeatDocumentSettings` data block) and their ranges can be easily translated into FDX markup.

FDX file has a block called `TagData`, which contains sections `TagCategories`, `TagDefinitions` and `Tags`. Each of the `TagCategories` has a long, hard-coded `id` in hex, ie. `01fc9642-84ff-4366-b37c-a3068dee57e8`. 

Tag definitions then refer to both `CatId` and their individual `Id`, for example:
`<TagDefinition CatId="01fc9642-84ff-4366-b37c-a3068dee57e8" Id="a12d3b1c-1053-1256-ac8d-1068b2d8a801" Label="Somebody" Number="1"/>`

This refers to the first tag, which is a character/cast tag. It's defined in the `Tags` section:  
````
    <Tag Number="1">
        <DefId>a12d3b1c-1053-1256-ac8d-1068b2d8a801</DefId>
    </Tag>
````
Here, the `DefId` contains the `Id` in the `TagDefinition`. It's a long, random hex. At first I thought it was a hash or hex representation of something, but no, it's just a random identifier. Tags can then reuse them, though I won't implement that.

The tag is then used inside the screenplay itself by adding `TagNumber` attribute to a `Text` element:  
`<Text TagNumber="1">Somebody</Text>`

I'll see into reworking the `FDXInterface` soon enough, but right now, I'm kind of overwhelmed by the string operations this will require. The original Writer code (which in turn is stolen from some Python implementation) should give some guidance on how to enumerate through ranges and converting them into XML.

**Comment by lmparppei** [comment]
`Line` object has been updated to contain `tags` array, which in turn has an `NSDictionary` with tag name and its range. `BeatTagging` class has a method for baking the values into the objects and converting the global ranges into local.

**Comment by lmparppei** [comment]
Implemented for the most part. Will be released for testing with 1.8 beta.
